dts: arm: st: add npgios to all stm32 gpio controllers

mrrosen opened the pull request:
For almost all STM32 GPIO controllers, the number of supported GPIO pins managed by a single controller is 16 (with some exceptions for fewer). However, the default for ngpios in the device tree bindings for gpio-controllers is 32; leading to inaccuracies in handling GPIO for these controllers, such as presenting too many GPIOs in the GPIO shell. This patch adds the ngpios property for all currently supported STM32 GPIO controllers.

mrrosen commented:
> Wouldn't it be better to override the default value of 32 and set it to 16 for STM32? (If it's possible to override a default?)
> We would avoid repeating the same line in practically all gpio nodes.

I agree but it didn't seem like it was possible to override defaults in the devicetree bindings and the default comes from gpio-controller.

FRASTM reviewed (commented):
could this be grouped in one change on **st,stm32-gpio.yaml**, like the following:
```
include:
  - name: base.yaml
  - name: gpio-controller.yaml
    property-blocklist:
      - ngpios

properties:
  ngpios:
    type: int
    default: 16

```

mrrosen commented:
> could this be grouped in one change on **st,stm32-gpio.yaml**, like the following:
> ```
> include:
>   - name: base.yaml
>   - name: gpio-controller.yaml
>     property-blocklist:
>       - ngpios
> 
> properties:
>   ngpios:
>     type: int
>     default: 16
> 
> ```
> 

That's an interesting way of doing it I didn't think of; but it does have the down side that fields (like the help text) are either lost for this binding or have to kept in sync with gpio-controller.yaml, right?

I do also wonder why then we wouldn't do this with all properties that are common then (like `gpio-controller` just defaulted true or `#gpip-cells` defaulted to 2 since those are duplicated for all gpio nodes too). I think it gets back to the discussion on that issue I linked where it didn't seem like it was desired to use the bindings as a replacement for actually setting values in the device tree but more just documenting things; and I'm not sure if Zephyr in general has changed it's mind on that since that thread is rather old.

erwango reviewed (commented):
Thanks @mrrosen. Sorry for not reviewing this earlier but I preferred to avoid tree wide changes before v3.7.0

I'm fine with updating ngpios for stm32, but then what about fixing the number in the stm32 compatible as done here for instance:
https://github.com/zephyrproject-rtos/zephyr/blob/e79ae3833e7fcf48a9799f1bc8359c3ca7facc3c/dts/bindings/gpio/nxp%2Cpcal6416a.yaml#L10-L13

erwango commented:
> I do also wonder why then we wouldn't do this with all properties that are common then (like gpio-controller just defaulted true or #gpip-cells defaulted to 2 since those are duplicated for all gpio nodes too). 

Because `ngpio` is already provided with a default value in `gpio-controller.yaml` which doesn't match STM32 value. Fixing it the same method feels as the reasonable approach.

erwango reviewed (changes requested):
Please use `default` method.

mrrosen committed:
dts: arm: st: correct npgios for all stm32 gpio controllers

For almost all STM32 GPIO controllers, the number of supported GPIO
pins managed by a single controller is 16 (with some exceptions for
fewer). However, the default for ngpios in the device tree bindings
for gpio-controllers is 32; leading to inaccuracies in handling GPIO
for these controllers, such as presenting too many GPIOs in the GPIO
shell. This patch redefines the default for ngpios for "st,stm32-gpio"
compatible devices to 16 and adds the correct ngpios for the few
exceptions Zephyr current supports.

Signed-off-by: Michael R Rosen <mrrosen@alumni.cmu.edu>

mrrosen commented:
@erwango @FRASTM @gautierg-st 

Ive changed the PR to align to the review discussion. Further reviewing the discussion in some of the places I linked; its does seems like this is more the Zephyr method to handling devicetree defaults (I would say that alot more properties can be defaulted as a result but those can be handled as they arise/need updating).

I would say however that while I dont really might be while bindings as a source of values for properties, I think its the fact that Zephyr's build doesnt seem to produce a final DTS with all the defaulted properties in place is rather unfortunate -- part of the idea using devicetree I felt was to let other external tools possible use it and with defaults used as they are here, that information is lost as no dts file ever actually has them. Or I am missing it, but I didnt see it in `zephyr/zephyr.dts` as I hoped I would...

erwango reviewed (dismissed):
> I think its the fact that Zephyr's build doesnt seem to produce a final DTS with all the defaulted properties in place is rather unfortunate -- part of the idea using devicetree I felt was to let other external tools possible use it and with defaults used as they are here, that information is lost as no dts file ever actually has them

That's indeed a relevant point, which however targets zephyr policy around defaults as a whole. Don't hesitate to raise an issue if you'd like to get this discussion to a wider audience.

gautierg-st commented on dts/bindings/gpio/st,stm32-gpio.yaml:
```diff
@@ -5,7 +5,11 @@ description: STM32 GPIO node
-include: [gpio-controller.yaml, base.yaml]
+include:
+  - name: gpio-controller.yaml
+    property-blocklist:
+      - ngpios
```
It doesn't seem necessary to do that. See the example provided by @erwango in nxp,pcal6416a.yaml, I don't see a property-blocklist (and it presumably works).

mrrosen replied:
Since that used `const`/`required`, it basically just added a `const` to the property binding. However, if there was a tree that used it (I actually dont see this driver in any board/shield), it would still need to be in the tree; its just anything other than `ngpios = <8>;` would be an error (missing it would be a violation of `required` and any other value a violation of `const: 8`)

henrikbrixandersen commented on dts/bindings/gpio/st,stm32-gpio.yaml:
```diff
@@ -17,6 +21,10 @@ properties:
 
+  ngpios:
+    type: int
+    default: 16
```
The rationale behind `default` values must _always_ be documented in the binding.

mrrosen commented:
> Thanks @mrrosen. Sorry for not reviewing this earlier but I preferred to avoid tree wide changes before v3.7.0
> 
> I'm fine with updating ngpios for stm32, but then what about fixing the number in the stm32 compatible as done here for instance:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/e79ae3833e7fcf48a9799f1bc8359c3ca7facc3c/dts/bindings/gpio/nxp%2Cpcal6416a.yaml#L10-L13

The `const` solution doesnt really work for a few reasons from my quick tests:
1. Its typically paired with `required` and if so, it must appear in the tree. If the goal is to not define it tree but rely on the value from the bindings, this doesnt work
2. Just defining `const` without `required` does not include the value in the tree, nor should it from the description of that key: https://docs.zephyrproject.org/latest/build/dts/bindings-syntax.html#const. Unlike `default`, `const` is more of just a constraint than a definition.
3. While very rare, there are STM32 GPIO controllers that do not implement all 16 IOs, at least as far as their documentation describes; so its better if for those exceptions, the value can be changed to the correct one for which defining by `const` does not allow

And I cannot define a `default` without blocking the property as it leads to an error processing the bindings (I did try just to see and it does not work). I think what was done here is the best use of a binding-defined value which avoids needing to define it in most trees as thats what everyone here was looking for.
